### PR TITLE
Fix for Device Events (OpenCL and SPIR-V events) when running Multi-threaded Execution plans

### PR DESCRIPTION
**Pull request by jjfumero**
#### Description

This PR fixes the issue of sharing the `eventPool` associated to a backend-context within the TornadoVM runtime. 

This PR also updates one of the methods to reset all internal state associated to one backend. 

```diff
- driver.getDefaultDevice().reset();
+ driver.getDefaultDevice().clean();
```

This resets the code cache and all events associated to a command queue. This method can be also reached from the TornadoExecutionPlan:

```java
executionPlan.resetDevice();
```

#### Problem description

When running Tornado execution plans from many different Java threads, there was the possibility of accessing an internal data structure from many threads. This internal data structure holds the pointers to the associated low-level events in a list. However, this list is maintained by the `<Backend>DeviceContext`, which must be shared for all threads running on the same device. 

This PR extends the previous work to achieve a single command queue/ or stream, per Java thread. 

#### Backend/s tested

Mark the backends affected by this PR.

- [X] OpenCL
- [X] PTX
- [X] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [X] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [X] No

#### How to test the new patch?

```bash
$ make
$ make tests
```

### CODE REVIEW

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoTestBase.java [hunk 1]
```diff
@@ -42,7 +42,7 @@ public void before() {
         for (int i = 0; i < TornadoRuntime.getTornadoRuntime().getNumBackends(); i++) {
             final TornadoBackend driver = TornadoRuntime.getTornadoRuntime().getBackend(i);
             for (int j = 0; j < driver.getDeviceCount(); j++) {
-                driver.getDevice(j).reset();
+                driver.getDefaultDevice().resetAll();
```
[the remark this answers is not in the document]
Comment (jjfumero) [reply tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoTestBase.java:45]: `resetAll` refers to all internal state. But I think this is still a bad name. I will change it to `cleanUp` , or something like this.

[the remark this answers is not in the document]
Comment (mikepapadim) [reply tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoTestBase.java:45]: I agree that `cleanUP` seems appropriate. 

Also, the cleanup above it should be on all the devices as we iterate on the driver list. right? not just on the default

[the remark this answers is not in the document]
Comment (jjfumero) [reply tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/common/TornadoTestBase.java]: I updated to `clean`. M.. it should be for all the devices. This was a mistake during the refactoring.

**Review by mairooni — APPROVED** [review]
LGTM

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/api/TestIO.java [hunk 2]
```diff
@@ -177,8 +176,6 @@ public void testCopyInWithDevice() {
         FloatArray arrayB = createAndInitializeArray(N);
         FloatArray arrayC = new FloatArray(N);
 
-        TornadoRuntime.getTornadoRuntime().getDefaultDevice().reset();
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/api/TestIO.java:180]: Any reason of this deletion? Is it deliberate?

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/fails/TestFails.java [hunk 3]
```diff
@@ -48,7 +48,7 @@ private void reset() {
         for (int i = 0; i < TornadoRuntime.getTornadoRuntime().getNumBackends(); i++) {
             final TornadoBackend driver = TornadoRuntime.getTornadoRuntime().getBackend(i);
             for (int j = 0; j < driver.getDeviceCount(); j++) {
-                driver.getDevice(j).reset();
+                driver.getDefaultDevice().clean();
```
Comment (stratika) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/fails/TestFails.java:51]: why did we select the default device for all iterations over the `driver.getDeviceCount()`?

Comment (stratika) [reply tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/fails/TestFails.java:51]: After looking at this unit-test, this method is called only from `test01` and it should run only for the default device. So, why do we iterate over the number of devices that exist in the driver of all backends?

File: tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVBackend.java [hunk 4]
```diff
@@ -230,8 +230,8 @@ public SPIRVDeviceContext getDeviceContext() {
         return this.context;
     }
 
-    public void reset() {
-        getDeviceContext().reset();
+    public void reset(long executionPlanId) {
```
Comment (stratika) [inline tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVBackend.java:233]: Why the `PTXBackend` does not have such a method?

File: tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/PTXDeviceContext.java [hunk 5]
```diff
@@ -152,6 +152,11 @@ public int getDriverIndex() {
         return TornadoRuntime.getTornadoRuntime().getBackendIndex(PTXBackendImpl.class);
     }
 
+    @Override
+    public Set<Long> getRegisteredPlanIds() {
+        return Set.of();
```
Comment (stratika) [inline tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/PTXDeviceContext.java:157]: Why does this class do not store the `executionIDs` similar to the other two (OCLDeviceContext, SPIRVDeviceContext)?

**Review by stratika — COMMENTED** [review]
In OpenCL, I get the following error:

```bash

tornado -ea  --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans"
WARNING: Using incubator modules: jdk.incubator.vector
Running thread t0Running thread t1Exception in thread "Thread-206" Exception in thread "Thread-205" uk.ac.manchester.tornado.api.exceptions.TornadoOutOfMemoryException: Unable to allocate 265814040 bytes of memory.
	at tornado.drivers.common@1.0.4-dev/uk.ac.manchester.tornado.drivers.common.TornadoBufferProvider.freeUnusedNativeBufferAndAssignRegion(TornadoBufferProvider.java:126)
	at tornado.drivers.common@1.0.4-dev/uk.ac.manchester.tornado.drivers.common.TornadoBufferProvider.getOrAllocateBufferWithSize(TornadoBufferProvider.java:153)
	at tornado.drivers.opencl@1.0.4-dev/uk.ac.manchester.tornado.drivers.opencl.mm.OCLMemorySegmentWrapper.allocate(OCLMemorySegmentWrapper.java:184)
	at tornado.drivers.opencl@1.0.4-dev/uk.ac.manchester.tornado.drivers.opencl.runtime.OCLTornadoDevice.newDeviceBufferAllocation(OCLTornadoDevice.java:577)
	at tornado.drivers.opencl@1.0.4-dev/uk.ac.manchester.tornado.drivers.opencl.runtime.OCLTornadoDevice.allocate(OCLTornadoDevice.java:590)
	at tornado.drivers.opencl@1.0.4-dev/uk.ac.manchester.tornado.drivers.opencl.runtime.OCLTornadoDevice.allocateObjects(OCLTornadoDevice.java:567)
	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.interpreter.TornadoVMInterpreter.executeAlloc(TornadoVMInterpreter.java:414)
	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.interpreter.TornadoVMInterpreter.execute(TornadoVMInterpreter.java:279)
	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.interpreter.TornadoVMInterpreter.execute(TornadoVMInterpreter.java:855)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.TornadoVM.executeInterpreterSingleThreaded(TornadoVM.java:125)
	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.TornadoVM.execute(TornadoVM.java:112)
	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph.scheduleInner(TornadoTaskGraph.java:859)
	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph.execute(TornadoTaskGraph.java:1366)
	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph.execute(TornadoTaskGraph.java:1378)
	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.TaskGraph.execute(TaskGraph.java:777)
	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.ImmutableTaskGraph.execute(ImmutableTaskGraph.java:49)
	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.TornadoExecutionPlan$TornadoExecutor.lambda$execute$0(TornadoExecutionPlan.java:406)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.TornadoExecutionPlan$TornadoExecutor.execute(TornadoExecutionPlan.java:406)
	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.TornadoExecutionPlan.execute(TornadoExecutionPlan.java:117)
	at tornado.unittests@1.0.4-dev/uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans.compute(TestMultiThreadedExecutionPlans.java:158)
	at tornado.unittests@1.0.4-dev/uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans.lambda$test04$7(TestMultiThreadedExecutionPlans.java:192)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

We need to add in the `tornado-test` script as parameters of the test the following flag: `-Dtornado.device.memory=4GB`.

If I run: 
```bash
tornado -ea  --jvm "-Xmx6g -Dtornado.device.memory=4GB -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans"
WARNING: Using incubator modules: jdk.incubator.vector
Running thread t0Running thread t1Test: class uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans
	Running test: test01                     ................  [PASS] 
	Running test: test02                     ................  [PASS] 
	Running test: test03                     ................  [PASS] 
	Running test: test04                     ................  [PASS]
```

**Review by mikepapadim — APPROVED** [review]
LGTM, works in popos

**Comment by jjfumero** [comment]
> In OpenCL, I get the following error:
> 
> ```shell
> tornado -ea  --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans"
> WARNING: Using incubator modules: jdk.incubator.vector
> Running thread t0Running thread t1Exception in thread "Thread-206" Exception in thread "Thread-205" uk.ac.manchester.tornado.api.exceptions.TornadoOutOfMemoryException: Unable to allocate 265814040 bytes of memory.
> 	at tornado.drivers.common@1.0.4-dev/uk.ac.manchester.tornado.drivers.common.TornadoBufferProvider.freeUnusedNativeBufferAndAssignRegion(TornadoBufferProvider.java:126)
> 	at tornado.drivers.common@1.0.4-dev/uk.ac.manchester.tornado.drivers.common.TornadoBufferProvider.getOrAllocateBufferWithSize(TornadoBufferProvider.java:153)
> 	at tornado.drivers.opencl@1.0.4-dev/uk.ac.manchester.tornado.drivers.opencl.mm.OCLMemorySegmentWrapper.allocate(OCLMemorySegmentWrapper.java:184)
> 	at tornado.drivers.opencl@1.0.4-dev/uk.ac.manchester.tornado.drivers.opencl.runtime.OCLTornadoDevice.newDeviceBufferAllocation(OCLTornadoDevice.java:577)
> 	at tornado.drivers.opencl@1.0.4-dev/uk.ac.manchester.tornado.drivers.opencl.runtime.OCLTornadoDevice.allocate(OCLTornadoDevice.java:590)
> 	at tornado.drivers.opencl@1.0.4-dev/uk.ac.manchester.tornado.drivers.opencl.runtime.OCLTornadoDevice.allocateObjects(OCLTornadoDevice.java:567)
> 	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.interpreter.TornadoVMInterpreter.executeAlloc(TornadoVMInterpreter.java:414)
> 	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.interpreter.TornadoVMInterpreter.execute(TornadoVMInterpreter.java:279)
> 	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.interpreter.TornadoVMInterpreter.execute(TornadoVMInterpreter.java:855)
> 	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
> 	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:762)
> 	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.TornadoVM.executeInterpreterSingleThreaded(TornadoVM.java:125)
> 	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.TornadoVM.execute(TornadoVM.java:112)
> 	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph.scheduleInner(TornadoTaskGraph.java:859)
> 	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph.execute(TornadoTaskGraph.java:1366)
> 	at tornado.runtime@1.0.4-dev/uk.ac.manchester.tornado.runtime.tasks.TornadoTaskGraph.execute(TornadoTaskGraph.java:1378)
> 	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.TaskGraph.execute(TaskGraph.java:777)
> 	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.ImmutableTaskGraph.execute(ImmutableTaskGraph.java:49)
> 	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.TornadoExecutionPlan$TornadoExecutor.lambda$execute$0(TornadoExecutionPlan.java:406)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> 	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.TornadoExecutionPlan$TornadoExecutor.execute(TornadoExecutionPlan.java:406)
> 	at tornado.api@1.0.4-dev/uk.ac.manchester.tornado.api.TornadoExecutionPlan.execute(TornadoExecutionPlan.java:117)
> 	at tornado.unittests@1.0.4-dev/uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans.compute(TestMultiThreadedExecutionPlans.java:158)
> 	at tornado.unittests@1.0.4-dev/uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans.lambda$test04$7(TestMultiThreadedExecutionPlans.java:192)
> 	at java.base/java.lang.Thread.run(Thread.java:1583)
> ```
> 
> We need to add in the `tornado-test` script as parameters of the test the following flag: `-Dtornado.device.memory=4GB`.
> 
> If I run:
> 
> ```shell
> tornado -ea  --jvm "-Xmx6g -Dtornado.device.memory=4GB -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans"
> WARNING: Using incubator modules: jdk.incubator.vector
> Running thread t0Running thread t1Test: class uk.ac.manchester.tornado.unittests.multithreaded.TestMultiThreadedExecutionPlans
> 	Running test: test01                     ................  [PASS] 
> 	Running test: test02                     ................  [PASS] 
> 	Running test: test03                     ................  [PASS] 
> 	Running test: test04                     ................  [PASS]
> ```

Yes, This is expected. For this test we need to increase the device heap size. It is already controlled in the `tornado-test` script

**Comment by jjfumero** [comment]
I should mention, this PR is connected to #389.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/api/TestIO.java (hunk 2 above)
Comment (jjfumero) [reply tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/api/TestIO.java:180]: There is no need for this call in this test

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/fails/TestFails.java (hunk 3 above)
Comment (jjfumero) [reply tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/fails/TestFails.java:51]: This was a mistake after replicating a change from another test. I will fix this.

File: tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVBackend.java (hunk 4 above)
Comment (jjfumero) [reply tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/SPIRVBackend.java:233]: This method was never triggered.

**Comment by jjfumero** [comment]
All comments adressed

**Review by stratika — APPROVED** [review]

**Pull request merged**